busbus/provider/gtfs.py

```diff
@@ -47,23 +72,136 @@ def __init__(self, provider, **data):
 
     @property
     def children(self):
-        return iter(self._provider._gtfs_stop_child_index.get(self.id, []))
+        return Queryable(self._provider._get_relation(self, 'children'))
+
+    @property
+    def _trips(self):
+        return Queryable(self._provider._get_relation(self, 'trips'))
 
 
 class GTFSRoute(busbus.Route):
 
     def __init__(self, provider, **data):
         if '_agency_id' in data:
-            data['agency'] = busbus.entity.LazyEntityProperty(
-                provider.get, busbus.Agency, data['_agency_id'])
+            data['agency'] = provider.get(busbus.Agency, data['_agency_id'])
         if '_type' in data:
             pass  # FIXME
 
         super(GTFSRoute, self).__init__(provider, **data)
 
 
-GTFS_FILENAME_MAP = {
-    'agency.txt': {
+class GTFSArrival(busbus.Arrival):
+    __derived__ = True
+
+    def __init__(self, provider, **data):
+        super(GTFSArrival, self).__init__(provider, **data)
+
+
+class GTFSService(busbus.entity.BaseEntity):
+    __attrs__ = ('id', 'days', 'start_date', 'end_date', 'added_dates',
+                 'removed_dates')
+
+    def __init__(self, provider, **data):
+        # Days of the week: 1 = Monday, 7 = Sunday (ISO 8601)
+        days = ('monday', 'tuesday', 'wednesday', 'thursday', 'friday',
+                'saturday', 'sunday')
+        data['days'] = list(i + 1 for i, day in enumerate(days)
+                            if data.get(day) == u'1')
+        for attr in ('start_date', 'end_date'):
+            if attr in data:
+                data[attr] = arrow.get(data[attr], 'YYYYMMDD').date()
+        data['added_dates'] = set()
+        data['removed_dates'] = set()
+
+        super(GTFSService, self).__init__(provider, **data)
+
+    @classmethod
+    def handle_calendar_dates(cls, provider, **data):
+        if not all(x in data for x in ('id', 'date', 'type')):
+            return
+        service = provider.get(cls, data['id'])
+        attr = {u'1': 'added_dates', u'2': 'removed_dates'}[data['type']]
+        getattr(service, attr).add(arrow.get(data['date'], 'YYYYMMDD').date())
+
+
+class GTFSTrip(busbus.entity.BaseEntity):
+    __attrs__ = ('id', 'route', 'service', 'headsign', 'short_name', 'block',
+                 'shape', 'accessible', 'bikes_ok')
+
+    def __init__(self, provider, **data):
+        data['route'] = provider.get(busbus.Route, data['_route_id'])
+        data['service'] = provider.get(GTFSService, data['_service_id'])
+        if '_block_id' in data:
+            pass  # FIXME
```

Comment:
There are a few FIXMEs in this vicinity. The underlying data types may be out of scope for this PR, though.